include/metaprogram.h

```diff
@@ -0,0 +1,72 @@
+/* Macros to aid with metaprogramming. */
+#ifndef METAPROGRAM_H
+#define METAPROGRAM_H
+
+/* Calls m0/m1/.../m8 depending on how many arguments are passed. */
+#define VARARG_8(m, ...) CAT(m, NARG_8(__VA_ARGS__))(__VA_ARGS__)
+
+/* Returns the number of arguments passed to it (up to 8). */
+#define NARG_8(...) NARG_8_(_, ##__VA_ARGS__, 8, 7, 6, 5, 4, 3, 2, 1, 0)
+#define NARG_8_(_, a, b, c, d, e, f, g, h, N, ...) N
+
+/* Expands 'a' and 'b' and then concatenates them. */
+#define CAT(a, b) CAT_(a, b)
+#define CAT_(a, b) a ## b
+
+/* Expands '__VA_ARGS__' and then stringizes them. */
+#define STR(...) STR_(__VA_ARGS__)
+#define STR_(...) #__VA_ARGS__
+
+/* Expands to the first/second/third/fourth argument. */
+#define FIRST(a, ...) a
+#define SECOND(a, b, ...) b
+#define THIRD(a, b, c, ...) c
+#define FOURTH(a, b, c, d, ...) d
+
+/* Simply returns __VA_ARGS__ */
+#define GET_ARGS(a, ...) __VA_ARGS__
```

Comment:
Is there a better name we could give to this? I don't think `GET_ARGS(arg0, arg1, arg2, arg3)` really communicates that it's `arg1, arg2, arg3` that gets returned.

Common names for this sort of operation are `TAIL` (which is paired with `HEAD`) or `REST` (which is paired with `FIRST`). But those usually operate on lists, and if you're not familiar with them in other languages I think they'd be just as opaque.

Maybe `DROP_FIRST`, or `EXCEPT_FIRST`, or something along those lines?

Reply:
Maybe `LIST_ARGS`? AFAIK we don't really have a "list" implementation in a traditional sense, so I think it's fine (otherwise in C++/C#/etc. it'd probably be less okay).

As for the selection, would it be easier just to specify a range? So providing `(0, 3)` would give you the first four and so on? You'd still need something like `DROP(X)` for managing items that "break sequence" but I think that's a good idea.

Reply:
The problem with `LIST_ARGS` is the same sort of problem as with `GET_ARGS`. It's the `ARGS` part. If you see `ARGS(a, b, c)` surely you'd think arguments, in which case you'd expect `a, b, c` to be the result. `ARGS` here _actually_ is derived from `__VA_ARGS__`, but that's really difficult to understand from the call site, because without looking at the definition you don't know which arguments are named and which are varargs (and further, I don't see why you'd expect everything but the first to be varargs without knowing the definition).

A slice-ish syntax doesn't sound unreasonable, feel free to suggest an implementation.

Reply:
I think I'll just drop this one instead, too tired for bikeshedding